ports/esp32/network_lan.c

```diff
@@ -0,0 +1,228 @@
+
+typedef struct _lan_if_obj_t {
+    mp_obj_base_t               base;
+    int                         if_id; // MUST BE FIRST
+    bool                        initialized;
+    bool                        active;
+    uint8_t                     mdc_pin;
+    uint8_t                     mdio_pin;
+    int8_t                      phy_power_pin;
+    uint8_t                     phy_addr;
+    uint8_t                     phy_type;
+    eth_phy_check_link_func     link_func;
+    eth_phy_power_enable_func   power_func;
+} lan_if_obj_t;
+
+const mp_obj_type_t lan_if_type;
+STATIC lan_if_obj_t lan_obj = {{&lan_if_type}, ESP_IF_ETH, false, false};
+
+STATIC void phy_power_enable(bool enable) {
+    lan_if_obj_t* self = MP_OBJ_TO_PTR(&lan_obj);
+
+    if (self->phy_power_pin != -1) {
+
+        if (!enable) {
+            /* Do the PHY-specific power_enable(false) function before powering down */
+            self->power_func(false);
+        }
+
+        gpio_pad_select_gpio(self->phy_power_pin);
+        gpio_set_direction(self->phy_power_pin,GPIO_MODE_OUTPUT);
+        if(enable == true) {
+            gpio_set_level(self->phy_power_pin, 1);
+        } else {
+            gpio_set_level(self->phy_power_pin, 0);
+        }
+
+        // Allow the power up/down to take effect, min 300us
+        vTaskDelay(1);
+
+        if (enable) {
+            /* Run the PHY-specific power on operations now the PHY has power */
+            self->power_func(true);
+        }
+    }
+}
+
+STATIC void init_lan_rmii() {
+    lan_if_obj_t* self = MP_OBJ_TO_PTR(&lan_obj);
+    phy_rmii_configure_data_interface_pins();
+    phy_rmii_smi_configure_pins(self->mdc_pin, self->mdio_pin);
+}
+
+STATIC void init_lan() {
+    lan_if_obj_t* self = MP_OBJ_TO_PTR(&lan_obj);
+    eth_config_t config;
+
+    switch (self->phy_type) {
+        case PHY_TLK110:
+            config = phy_tlk110_default_ethernet_config; 
+            break;
+        case PHY_LAN8720:
+            config = phy_lan8720_default_ethernet_config;
+            break;
+    }
+    
+    self->link_func = config.phy_check_link;
+
+    // Replace default power func with our own
+    self->power_func = config.phy_power_enable;
+    config.phy_power_enable = phy_power_enable;
+
+    config.phy_addr = self->phy_addr;
+    config.gpio_config = init_lan_rmii;
+    config.tcpip_input = tcpip_adapter_eth_input;
+
+    if (esp_eth_init(&config) == ESP_OK) {
+        esp_eth_enable();
+        self->active = true;
+    } else {
+        mp_raise_msg(&mp_type_OSError, "esp_eth_init() failed");
+    }
+}
+
+STATIC mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    lan_if_obj_t* self = MP_OBJ_TO_PTR(&lan_obj);
+
+    if (self->initialized) {
+        mp_raise_msg(&mp_type_OSError, "ethernet already initialized");
+    }
+
+    enum { ARG_id, ARG_mdc, ARG_mdio, ARG_power, ARG_phy_addr, ARG_phy_type };
+
+    uint8_t default_pins[] = {23, 18, 17}; // mdc, mdio, power
```

Comment:
Is it sensible to have default pins?  Maybe it would be simpler for the implementation here, and cleaner in the uPy scripts, if the user is forced to specify these pins?

Reply:
You're probably right.  These are the _de facto_ default pins, but they're also the only ones changeable.

I suppose the same argument could be made for PHY address and PHY type?

Thoughts?

Reply:
Nudge

Reply:
I think everything should be required here, ie no defaults.  It's easy to make them defaults later without breaking code, but not the other way around.